### PR TITLE
[koa] add koa constructor params support

### PR DESCRIPTION
**Pull request by vnoder**
add koa constructor params type support.

Please fill in this template.

- [ ] Use a meaningful title for the pull request. Include the name of the package modified.
- [ ] Test the change in your own code. (Compile and run.)
- [ ] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [ ] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [ ] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [ ] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If adding a new definition:
- [ ] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [ ] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [ ] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [ ] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [ ] `tslint.json` [should contain](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#linter-tslintjson) `{ "extends": "dtslint/dt.json" }`, and no additional rules.
- [ ] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

If removing a declaration:
- [ ] If a package was never on Definitely Typed, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [ ] Delete the package's directory.
- [ ] Add it to `notNeededPackages.json`.

### CODE REVIEW

File: types/koa/index.d.ts [hunk 1]
```diff
@@ -456,7 +456,14 @@ declare class Application<
     silent: boolean;
     keys: Keygrip | string[];
 
-    constructor();
+    constructor(options?: {
```
Comment (pkuczynski) [inline types/koa/index.d.ts:459]: Koa offers some nice jsdoc https://github.com/koajs/koa/blob/master/lib/application.js#L39

Maybe it's worth adding it here too?

**Review by harryparkdotio — APPROVED** [review]

File: types/koa/index.d.ts (hunk 1 above)
Comment (vnoder) [reply types/koa/index.d.ts:471]: Year, I add it together

File: types/koa/index.d.ts [hunk 2]
```diff
@@ -456,7 +457,25 @@ declare class Application<
     silent: boolean;
     keys: Keygrip | string[];
 
-    constructor();
+    /**
+     *
+     * @param {object} [options] Application optionsJake, 2 years ago: • feat: accept options in the Application const…
```
Comment (pkuczynski) [inline types/koa/index.d.ts:462]: @vnoder I think there is accidentally added text here, if I compare it with https://github.com/koajs/koa/blob/698ce0afbfac6480400625729a4b8fc4b4203fdc/lib/application.js#L39

**Review by pkuczynski — CHANGES_REQUESTED** [review]

**Review by pkuczynski — APPROVED** [review]

**Comment by fwh1990** [comment]
@vnoder why define proxyIpHeader to `boolean`, exactly, it can be `undefined | string`
see here https://github.com/koajs/koa/blob/master/lib/application.js#L54
and here https://github.com/koajs/koa/blob/master/lib/request.js#L435

**Pull request merged**